targets/TARGET_Silicon_Labs/TARGET_EFM32/can_api.c

```diff
@@ -0,0 +1,310 @@
+                   CanInit.phaseBufferSegment1,
+                   CanInit.phaseBufferSegment2,
+                   CanInit.synchronisationJumpWidth);
+}
+
+int can_write(can_t *obj, CAN_Message msg, int cc)
+{
+    CAN_MessageObject_TypeDef message;
+
+    message.id = msg.id;
+    message.msgNum = 1;
+
+    if (msg.format == 0) message.extended = false;
+    else message.extended = true;
+
+    message.dlc = msg.len;
+
+    for (int i = 0; i < message.dlc; ++i) {
+        message.data[i] = (uint32_t)msg.data[i];
+    }
+
+    CAN_SendMessage(obj->instance, CAN_TX_IF, &message, true);
+
+    return 1;
+}
+
+int can_read(can_t *obj, CAN_Message *msg, int handle)
+{
+	CAN_MessageObject_TypeDef receiver;
+
+	if (CAN_HasNewdata(obj->instance)) {
+
+		receiver.msgNum = 2;
+
+		CAN_ReadMessage(obj->instance, CAN_RX_IF, &receiver);
+
+		msg->id = receiver.id;
+		msg->len = receiver.dlc;
+
+		for (int i = 0; i < receiver.dlc; ++i) {
+			msg->data[i] = (unsigned char)receiver.data[i];
+		}
+
+	    return 1;
+	}
+
+	return 0;
+}
+
+void can_reset(can_t *obj)
+{
+    CAN_Reset(obj->instance);
+}
+
+unsigned char can_rderror(can_t *obj)
+{
+	return ((obj->instance->ERRCNT>>_CAN_ERRCNT_REC_SHIFT)&0xFF);
+}
+
+unsigned char can_tderror(can_t *obj)
+{
+	return ((obj->instance->ERRCNT)&0xFF);
+}
+
+void can_monitor(can_t *obj, int silent)
+{
+	CanMode mode = MODE_NORMAL;
+
+    if (silent) {
+    	mode = MODE_SILENT;
+    }
+
+    CAN_SetMode(obj->instance, mode);
+}
+
+int can_mode(can_t *obj, CanMode mode)
+{
+	CAN_SetMode(obj->instance, mode);
+	return 1;
+}
+
+int can_filter(can_t *obj, uint32_t id, uint32_t mask, CANFormat format, int32_t handle)
+{
+	CAN_MessageObject_TypeDef message;
+
+	message.msgNum = 2;
+	message.id = id;
+	message.mask = mask;
+
+	if (format == CANStandard) {
+	  message.extended = false;
+	  message.extendedMask = false;
+	} else {
+	  message.extended = true;
+	  message.extendedMask = true;
+	}
+
+	CAN_SetIdAndFilter(obj->instance, CAN_RX_IF, true, &message, true);
+
+	return handle;
+}
+
+void can_irq_set(can_t *obj, CanIrqType type, uint32_t enable)
+{
+	uint32_t ier = _CAN_IF0IEN_MESSAGE_MASK;
+
+    if (enable) {
+    	CAN_MessageIntEnable(obj->instance, ier);
+    } else {
+    	CAN_MessageIntDisable(obj->instance, ier);
+    }
+
+	switch ((CANName)obj->instance) {
+        case CAN_0:
+        	NVIC_SetVector(CAN0_IRQn, CAN0_IRQHandler);
+        	NVIC_EnableIRQ(CAN0_IRQn);
+            break;
+        case CAN_1:
+        	NVIC_SetVector(CAN1_IRQn, CAN1_IRQHandler);
+        	NVIC_EnableIRQ(CAN1_IRQn);
+            break;
+    }
+}
+
+static void can_irq(CANName name, int id)
+{
+	CAN_TypeDef *can;
+	can = (CAN_TypeDef *)name;
+
+	if (can->STATUS & CAN_STATUS_RXOK) {
+		irq_handler(can_irq_ids[id], IRQ_RX);
+	    CAN_MessageIntClear(can, 0xFFFFFFFF);
+	}
+
+	if (can->STATUS & CAN_STATUS_TXOK) {
+		irq_handler(can_irq_ids[id], IRQ_TX);
+	    CAN_MessageIntClear(can, 0xFFFFFFFF);
+	}
+}
+
+void CAN0_IRQHandler(void)
```

Comment:
Please wrap in respective peripheral base definition, as this shouldn't be there if there is no (for example) CAN1.